app/utils/hash_utils.rb

```diff
@@ -144,4 +144,39 @@ def flatten(h)
 
     acc.call(nil, h)
   end
+
+  # Recursively walks through nested hash and performs `map` operation.
+  #
+  # The tree is traversed in pre-order manner.
+  #
+  # In each node, calls the block with two arguments: key and value.
+  # The block needs to return a tuple of [key, value].
+  #
+  # Example (double all values):
+  #
+  # deep_map(a: { b: { c: 1}, d: [{ e: 1, f: 2 }]}) { |k, v|
+  #   [k, v * 2]
+  # }
+  #
+  #
+  # Example (stringify keys):
+  #
+  # deep_map(a: 1, b: 2) { |k, v|
+  #   [k.to_s, v]
+  # }
+  #
+  # Unlike Ruby's Hash#map, this method returns a Hash, not an Array.
+  #
+  def deep_map(obj, &block)
```

Comment:
I don't know what is the actual definition of "walking", but to me it sounds more generic than this operation, which is "mapping". To me "walking" sounds like iterating all the nodes, but doesn't specify what the operation is (e.g. replacing all the `.map` calls in this function with `.each`). But again, I'm not sure what's the correct definition. Do you know?